Sorbet/README.md

```diff
@@ -0,0 +1,113 @@
+
+These are standard Ruby classes, chosen for their simplicity and immutability, making them suitable for testing type narrowing in Sorbet.
+
+> Q. What container types does this implementation use (for objects, tuples, etc)? Why?
+
+* Hash types for objects: `{ a: T.untyped }` or `{ a: T.any(String, Integer) }`
+* Array types for tuples: `[T.untyped, T.untyped]` or `[Integer, T.any(String, Integer)]`
+* Union types for sized tuples: `T.any([Integer, Integer], [String, String, String])`
+
+Hashes represent key-value objects, common in Ruby. Arrays serve as tuples, with fixed or variable lengths. Union types model sized tuples by distinguishing lengths (e.g., 2 vs. 3 elements). These types are immutable in the context of the benchmark, ensuring type soundness.
+
+## Type Narrowing
+
+> Q. How do simple type tests work in this language?
+
+Sorbet uses Ruby’s `is_a?` method for type tests, e.g., `x.is_a?(String)`. These refine types in conditional branches when Sorbet can prove the type is narrowed.
+
+Example: `if x.is_a?(String); x.length; else; 0; end`
+
+<https://sorbet.org/docs/flow-sensitive>
+
+> Q. Are there other forms of type test? If so, explain.
+
+* `T.assert_type!(x, T)`: Asserts `x` has type `T`, raising an error if not, and refining the type in the current scope.
+* `case` statements with type checks: Sorbet can narrow types based on `when` clauses using `is_a?`.
+
+<https://sorbet.org/docs/type-assertions>
+
+> Q. How do type casts work in this language?
+
+`T.let(x, T)` and `T.cast(x, T)` assign type `T` to `x`. `T.let` is used for local type annotations, while `T.cast` inserts runtime checks (in strict mode) to enforce the type. Both update the type environment statically.
+
+Example: `T.cast(x, String)` ensures `x` is treated as `String`.
+
+<https://sorbet.org/docs/type-assertions>
+
+> Q. What is the syntax for a symmetric (2-way) type-narrowing predicate?
+
+Sorbet does not support type-narrowing predicates (symmetric or otherwise). Functions returning `T::Boolean` (e.g., `x.is_a?(String)`) do not refine the argument’s type in the caller’s scope.
+
+Example: A function like `sig { params(x: T.any(String, Integer)).returns(T::Boolean) }` cannot refine `x` to `String` or `Integer` based on the return value.
+
+<https://sorbet.org/docs/from-typescript>
+
+> Q. If the language supports other type-narrowing predicates, describe them below.
+
+Sorbet lacks support for type-narrowing predicates, including one-way or checked predicates. The benchmark implements these as stubs with dummy returns (`true`, `0`) for success cases and type errors (e.g., `x.is_nan`) for failure cases to highlight this limitation.
+
+## Benchmark Details
+
+> Q. Are any benchmarks inexpressible? Why?
+
+No, all benchmarks are expressible in Sorbet. However, predicate-related benchmarks (`predicate_2way`, `predicate_1way`, `predicate_checked`) are implemented as stubs due to Sorbet’s lack of type predicate support.
```

Comment:
Reword this. (1) predicates are not expressible. (2) all other benchmarks are expressible.